### PR TITLE
Handle diff3-style merge correctly

### PR DESCRIPTION
**Pull request by katemihalikova**
Fixes pprice/vscode-better-merge#23.

I'm using `merge.conflictStyle = diff3` git setting, but the style (_Current/Common ancestors/Incoming_) is not supported in VS Code. In the new merge extension (#27150), the _Common ancestors_ block is merged together with the _Current_ block. This PR adds that support. As `diff3` is a superset of the default git conflict style, I've modified the original algorithm to support both styles. The same fix was also proposed by @spiffytech in the original issue.

### CODE REVIEW

**Comment by sandy081** [comment]
@kieferrm Not sure who owns the merge extension, hence assigning this to you. Please re-assign it to the right person.

Thanks.

File: extensions/merge-conflict/src/mergeConflictParser.ts [hunk 1]
```diff
@@ -49,7 +51,11 @@ export class MergeConflictParser {
 				}
 
 				// Create a new conflict starting at this line
-				currentConflict = { startHeader: line };
+				currentConflict = { startHeader: line, commonAncestors: null };
+			}
+			// Are we within a conflict block and is this a common ancestors marker? |||||||
+			else if (currentConflict && line.text.startsWith(commonAncestorsMarker)) {
+				currentConflict.commonAncestors = line;
```
Comment (pprice) [inline extensions/merge-conflict/src/mergeConflictParser.ts]: If, for some reason, we ended up with:

```
<<<<<<< HEAD
Foo
||||||| common
foo
||||||| common
fooo
=======
blah
>>>>>>> incoming
```

We will end up extending the "current" content over the first merge ancestor block. If we encounter two blocks of `|||||||` I think there are two thing we could do: 

 - Favorable: Don't overwrite `currentConflict.commonAncestors`, just leave it with the first occurrence, this way the current content block will span the correct range. (e.g. `else if (currentConflict && !currentConflict.commonAncestors && line.text.startsWith(commonAncestorsMarker))`
 - Break out of scanning (like we do if we encounter a start marker before an end marker), so we don't provide any decoration if we are unsure what to do. 

FWIW I think this is also true of existing `=======` scanning...

**Review by pprice — COMMENTED** [review]
🎉 Thank you @katemihalikova LGTM! @chrmarti / @kieferrm can give final approval. 

I tested Kate's fork locally, and everything appears to be working great.

File: extensions/merge-conflict/src/mergeConflictParser.ts [hunk 2]
```diff
@@ -7,11 +7,13 @@ import * as interfaces from './interfaces';
 import { DocumentMergeConflict } from './documentMergeConflict';
 
 const startHeaderMarker = '<<<<<<< ';
+const commonAncestorsMarker = '||||||| ';
 const splitterMarker = '=======';
 const endFooterMarker = '>>>>>>> ';
 
 interface IScanMergedConflict {
 	startHeader: vscode.TextLine;
+	commonAncestors: vscode.TextLine | null;
```
Comment (katemihalikova) [inline extensions/merge-conflict/src/mergeConflictParser.ts]: - [x] Marked as optional

File: extensions/merge-conflict/src/mergeConflictParser.ts [hunk 3]
```diff
@@ -84,6 +90,12 @@ export class MergeConflictParser {
 			return null;
 		}
 
+		let tokenAfterCurrentBlock: vscode.TextLine = scanned.splitter;
+
+		if (scanned.commonAncestors !== null) {
+			tokenAfterCurrentBlock = scanned.commonAncestors;
+		}
+
```
Comment (katemihalikova) [inline extensions/merge-conflict/src/mergeConflictParser.ts]: - [x] Condensed

File: extensions/merge-conflict/src/mergeConflictParser.ts (hunk 1 above)
Comment (katemihalikova) [inline extensions/merge-conflict/src/mergeConflictParser.ts]: - [x] Incorporated the first solution for both markers, now we are using just the first token of each type, just like TypeScript does.

**Comment by katemihalikova** [comment]
I've resolved the conflict.

**Comment by chrmarti** [comment]
Thanks for your contribution @katemihalikova.

I will add the workbench colors on top. One corner case is with multiple common ancestors where these have only a single header decoration. With some background colors applied this looks like:
screen shot 2017-06-07 at 9 49 33 am

Would adding one header and one content decoration for each common ancestor improve on this?

**Comment by katemihalikova** [comment]
Yeah that makes sense, I'll add it.

**Comment by katemihalikova** [comment]
@chrmarti The last commit added support for multiple common ancestors blocks:

image

(background colors are not included)

**Comment by chrmarti** [comment]
Thanks @katemihalikova !

**Pull request merged**